### PR TITLE
feat: Custom color schemes support

### PR DESCRIPTION
**Pull request by amitmiran137**
### SUMMARY
Today colors schemes are loaded from @superset-ui/core 
we should have support for individual deployments to inject their own sets of color scheme
the mechanism was copied from setup plugins logic

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: docs/src/pages/docs/frequently-asked-questions-page.mdx [hunk 1]
```diff
@@ -232,6 +232,28 @@ Metadata attribute using the `label_colors` key.
 }
 ```
 
+### Is there a way to add custom color schemes?
+
+It is possible to override [setupColorsExtra.ts](/src/setup/setupColorsExtra.ts) to register new color schemes
+
+```
+import CategoricalScheme from '../../CategoricalScheme'; // Can use existing scheme type or create new types
```
Comment (ktmud) [inline docs/src/pages/docs/frequently-asked-questions-page.mdx:240]: Should this be `import { CategoricalScheme } from '@superset-ui/core';`

**Review by ktmud — APPROVED** [review]
LGTM with one nit.

I'm also wondering whether we should consolidate all setup `extra`s to a single file and provide a somewhat more formalized API...

File: superset-frontend/src/setup/setupColors.js [hunk 2]
```diff
@@ -49,4 +50,6 @@ export default function setupColors() {
     });
   });
   sequentialSchemeRegistry.setDefaultKey('superset_seq_1');
+
+  setupColorsExtra();
```
Comment (mistercrunch) [inline superset-frontend/src/setup/setupColors.js:54]: Should we pass the context here as in `setupColorsExtra({sequentialSchemeRegistry, categoricalSchemeRegistry});` or similar?

I'm also wondering if we should have a central file for all the configuration hooks, similar to `superset/config.py` on the python side?

Comment (amitmiran137) [reply superset-frontend/src/setup/setupColors.js]: hi @mistercrunch I really liked your idea and added another commit  supporting color schemes from the config.py file

File: docs/src/pages/docs/frequently-asked-questions-page.mdx [hunk 3]
```diff
@@ -232,6 +232,29 @@ Metadata attribute using the `label_colors` key.
...
+```
+SETUP_EXTRA = {"COLOR_SCHEMES": [
+    {
+        "id": 'myCustomColors',
+        "schemeType": 'Categorical', //Categorical or Sequential
+        "description": '',
+        "label": 'My custom Colors',
+        "colors": ['#6BD3B3', '#6BD3B3', '#6BD3B3', '#66CBE2']
+    }]
+}
```
Comment (villebro) [inline docs/src/pages/docs/frequently-asked-questions-page.mdx:248]: `SETUP_EXTRA` feels ambiguous, and will be difficult/disruptive to change once introduced. Perhaps break this up into `EXTRA_COLOR_SCHEMES` or even `EXTRA_CATEGORICAL_COLOR_SCHEMES` and `EXTRA_SEQUENTIAL_COLOR_SCEHEMES` would make sense?

File: docs/src/pages/docs/frequently-asked-questions-page.mdx [hunk 4]
```diff
@@ -232,6 +232,29 @@ Metadata attribute using the `label_colors` key.
...
+.
+.
+.
+schemes.forEach(scheme => {
+    categoricalSchemeRegistry.registerValue(scheme.id, scheme);
+});
```
Comment (villebro) [inline docs/src/pages/docs/frequently-asked-questions-page.mdx:254]: Is this needed here? The FAQ should probably explain that the change should be done in `superset_config.py`.

File: superset-frontend/src/preamble.ts [hunk 5]
```diff
@@ -20,6 +20,7 @@ import { setConfig as setHotLoaderConfig } from 'react-hot-loader';
 import 'abortcontroller-polyfill/dist/abortcontroller-polyfill-only';
 import moment from 'moment';
 import { configure } from '@superset-ui/core';
+import ColorScheme from '@superset-ui/core/lib/color/ColorScheme';
```
Comment (villebro) [inline superset-frontend/src/preamble.ts:23]: We should probably export this (and other related ones) in `@superset-ui/core`.

File: superset-frontend/src/preamble.ts [hunk 6]
```diff
@@ -49,7 +52,17 @@ if (typeof window !== 'undefined') {
...
+
+if (
+  bootstrapData.common &&
+  bootstrapData.common.setup_extra &&
+  bootstrapData.common.setup_extra.COLOR_SCHEMES
+) {
+  color_schemas = bootstrapData.common.setup_extra.COLOR_SCHEMES;
+}
+
+setupColors(color_schemas);
```
Comment (villebro) [inline superset-frontend/src/preamble.ts:65]: Nit: 
```suggestion
let color_schemes: ColorScheme[] = [];

if (
  bootstrapData.common &&
  bootstrapData.common.setup_extra &&
  bootstrapData.common.setup_extra.COLOR_SCHEMES
) {
  color_schemes = bootstrapData.common.setup_extra.COLOR_SCHEMES;
}

setupColors(color_schemes);
```

**Review by villebro — COMMENTED** [review]
A few comments

File: docs/src/pages/docs/frequently-asked-questions-page.mdx (hunk 3 above)
Comment (amitmiran137) [reply docs/src/pages/docs/frequently-asked-questions-page.mdx]: makes sense. added changes accordingly

File: docs/src/pages/docs/frequently-asked-questions-page.mdx [hunk 7]
```diff
@@ -232,6 +232,30 @@ Metadata attribute using the `label_colors` key.
...
+    }]
+
+EXTRA_SEQUENTIAL_COLOR_SCHEMES =  [
+    {
+        "id": 'myCustomColors',
+        "description": '',
+        "isDiverging": true
+        "label": 'My custom Colors',
+        "colors": ['#6BD3B3', '#6BD3B3', '#66CBE2']
+    }]
```
Comment (villebro) [inline docs/src/pages/docs/frequently-asked-questions-page.mdx:255]: I would rather see this example in `config.py` with a reference to it in the FAQ. Also a minor nit, but would be nice to have the schemes having real colors and not repeating `#6BD3B3`, also with the name something like "My custom warm to hot" or similar.

File: superset/config.py [hunk 8]
```diff
@@ -368,6 +368,14 @@ def _try_json_readsha(  # pylint: disable=unused-argument
 #     return feature_flags_dict
 GET_FEATURE_FLAGS_FUNC: Optional[Callable[[Dict[str, bool]], Dict[str, bool]]] = None
 
+# EXTRA_CATEGORICAL_COLOR_SCHEMES is used for adding custom categorical color schemes
+# This is merely a default
+EXTRA_CATEGORICAL_COLOR_SCHEMES: List[Dict[str, Any]] = []
+
+# EXTRA_SEQUENTIAL_COLOR_SCHEMES is used for adding custom sequential color schemes
+# This is merely a default
+EXTRA_SEQUENTIAL_COLOR_SCHEMES: List[Dict[str, Any]] = []
```
Comment (villebro) [inline superset/config.py:377]: See the comment in FAQ

**Review by villebro — COMMENTED** [review]
Would you be ok with moving the example to `config.py`? LGTM after that.

**Review by villebro — APPROVED** [review]

**Pull request merged**